### PR TITLE
Site assembler - Categories revamp and cleanup feature flags

### PR DESCRIPTION
**Pull request by miksansegundo**
Related to https://github.com/Automattic/wp-calypso/pull/72144

## Proposed Changes

The UX is ready for testing. The code is going to be organized in smaller PRs. First the cleanup and then new features.

**Add and replace patterns from categories**

https://user-images.githubusercontent.com/1881481/223174881-5d7a0db5-38bf-489c-b5eb-cd398ddb7dba.mov

https://user-images.githubusercontent.com/1881481/223177086-702a2d0d-69df-46e7-bdf1-562faca2d4e0.mov

**More changes**

* Restyling
* Refactored components
* Removed old feature flags: `pattern-assembler/sidebar-revamp`, `pattern-assembler/client-side-render`, `pattern-assembler/write-flow`
* Added a new flag `pattern-assembler/categories`
* Added categories
* Added animation hooks
* Added a snackbar notice when section patterns are added
* Changed the copy on the Replace pattern screen
* Changed the copy on the Add patterns screen

**TODOs**
- **Remove horizontal padding from all pattern previews**
- **Move the focus to the list of patterns after they load - A11y**
- **Allow focus on the site preview iframe to scroll up/down or provide controls - A11y**
- **Get feedback on the copy to make the screens clearer**
- **Verify that all Trac metrics work**
- Add a snackbar notice when section patterns are replaced?
- Add a snackbar notice when header and footer patterns are added/replaced?
- Add new metrics?

## Testing Instructions

* Use the Calypso live link on the first comment
* Create a new site and click "Start designing" from the bottom of the design picker 
* Choose a header and a footer
* Create your homepage 
* Add patterns 
* Open categories
* Replace, order and delete patterns
* Test Trac metrics
* Test accessibility

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by davewhitley** [comment]
I'm wondering about the speed of the panel animation. On Safari at least, there is a noticeable slowdown that I don't think we can fix by changing the animation timing. I think it might be the properties we are animating (`left` and/or `opacity`) that are causing it to be sluggish. Could we use other techniques, maybe `translate`? Also, it could be that we are loading all of those pattern iframes at the same time. Could we prioritize the panel animation first, then the pattern loading?

Here's a smooth example:

https://user-images.githubusercontent.com/618551/223206960-ff15e801-f0a8-4982-98e2-8884b9acea8e.mp4

Let's use Gray 0 for the background of the panel, should compute to `#F6F7F7`.

> Add a snackbar notice when section patterns are replaced?

I think yes, and it should say "replaced" instead of "added".

> Add a snackbar notice when header and footer patterns are added/replaced?

No, it didn't feel right to me.

> Get feedback on the copy to make the screens clearer

Everything looked good to me but you might make the Replace screen a bit better with the secondary text:

"Replace the selected pattern by choosing from the list of categories below."

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-layout.tsx [hunk 1]
```diff
@@ -24,125 +23,22 @@ type PatternLayoutProps = {
...
-	onAddFooter = noop,
-	onReplaceFooter = noop,
-	onDeleteFooter = noop,
 }: PatternLayoutProps ) => {
 	const translate = useTranslate();
 	const [ isPopoverVisible, setIsPopoverVisible ] = useState( false );
 	const addButtonContainerRef = useRef< HTMLDivElement | null >( null );
-	const isSidebarRevampEnabled = isEnabled( 'pattern-assembler/sidebar-revamp' );
-
-	if ( ! isSidebarRevampEnabled ) {
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-layout.tsx:51]: Note: The feature flag is removed by https://github.com/Automattic/wp-calypso/pull/74149. We can ignore it from this PR

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-layout.tsx [hunk 2]
```diff
@@ -187,9 +83,8 @@ const PatternLayout = ( {
 					context={ addButtonContainerRef.current }
 					isVisible={ isPopoverVisible }
 					position="right"
-					focusOnShow
 				>
-					{ translate( 'Add your next pattern' ) }
+					{ translate( 'Add patterns' ) }
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-layout.tsx:87]: It's finished via https://github.com/Automattic/wp-calypso/pull/74104

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/style.scss [hunk 3]
```diff
@@ -23,24 +31,26 @@ $font-family: "SF Pro Text", $sans;
...
 		flex: 1;
 		outline: none;
+		position: relative;
 	}
 
 	.pattern-assembler__sidebar {
 		display: flex;
 		flex-direction: column;
-		width: 284px;
+		width: 343px;
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/style.scss:47]: This should be wrapped inside a class so it could be controlled by a feature flag.

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/screen-category-list.tsx [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+				}
+			/>
+			<div className="screen-container__body screen-category-list__body">
+				{ categoriesWithPatterns.map( ( { name, label, description } ) => {
+					const isOpen = categorySelected === name;
+					return (
+						<Button
+							key={ name }
+							className={ classNames( 'screen-category-list__category-button', {
+								'--is-open': isOpen,
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/screen-category-list.tsx:59]: We discussed this before. Let's change it to `screen-category-list__category-button--is-open`

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/screen-category-list.scss [hunk 5]
```diff
@@ -0,0 +1,30 @@
+
+
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/screen-category-list.scss:2]: Could we remove these empty lines?

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/index.tsx [hunk 6]
```diff
@@ -470,10 +516,29 @@ const PatternAssembler: Step = ( { navigation, flow, stepName } ) => {
...
+					placeholder: null,
+					header,
+					sections,
+					footer,
+					activePosition,
+				} }
+			/>
+			{ isEnabled( 'pattern-assembler/categories' ) && (
+				<AsyncLoad
+					require="./pattern-list"
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/index.tsx:531]: I think the new pattern panel should be controlled inside `ScreenCategoryList`. We can set the `overflow-x: visible` to `.pattern-assembler__sidebar`

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-list.tsx [hunk 7]
```diff
@@ -1,57 +1,73 @@
...
-		</>
+		<AnimateWithFeatures featureName="domAnimation">
+			{ ( m: any ) =>
+				openPatternList ? (
+					<m.div
+						key="pattern-list"
+						className="pattern-list__wrapper"
+						layoutScroll
+						initial={ { opacity: 0.2, left: 250 } }
+						animate={ { opacity: 1, left: 343 } }
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-list.tsx:52]: When the panel is controlled by `ScreenCategoryList`, we can change to use `transform: translateX(-100%)`.

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-list.scss [hunk 8]
```diff
@@ -1,26 +1,14 @@
...
-		position: absolute;
-		left: 0;
-		right: 0;
-		top: 0;
-		bottom: 0;
-	}
+.pattern-list__wrapper {
+	position: absolute;
+	top: -60px;
+	width: 343px;
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-list.scss:4]: When the panel is controlled by `ScreenCategoryList`, we can change to use `width: 100%`.

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-selector.tsx [hunk 9]
```diff
@@ -1,80 +1,67 @@
...
-		emptyPatternText,
-		activeClassName: 'pattern-selector__block-list--selected-pattern',
-		onSelect,
-	};
 
 	return (
 		<div className="pattern-selector">
-			{ title && (
-				<div className="pattern-selector__header">
+			{ ! isEnabled( 'pattern-assembler/categories' ) && showHeader && (
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-selector.tsx:34]: Currently, the Pattern Selector is focusing on showing the pattern list and it could be re-used by `ScreenHeader`, `ScreenFooter`, and `ScreenPatternList`, so I don't prefer to add a navigation header here. Instead, its parent has to add a header and footer on their own.

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-selector.tsx [hunk 10]
```diff
@@ -1,80 +1,67 @@
...
+							shownPatterns,
+							selectedPattern,
+							emptyPatternText,
+							activeClassName: 'pattern-selector__block-list--selected-pattern',
+							onSelect,
+						} }
+					/>
+				</div>
+			</div>
+			{ ! isEnabled( 'pattern-assembler/categories' ) && showDoneButton && (
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-selector.tsx:53]: Same but this has been moved out from the pattern selector

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/index.tsx [hunk 11]
```diff
@@ -171,6 +171,39 @@ const PatternAssembler: Step = ( { navigation, flow, stepName } ) => {
 			} as DesignRecipe,
 		} as Design );
 
+	// @TODO: handle the state in a new component <Notices>
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/index.tsx:174]: I think all of the notices should be separated into another PR as it's an enhancement and not a blocker for adding categories

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/index.tsx [hunk 12]
```diff
@@ -239,13 +273,26 @@ const PatternAssembler: Step = ( { navigation, flow, stepName } ) => {
 		updateActivePatternPosition( position - 1 );
 	};
 
-	const onSelect = ( type: string, selectedPattern: Pattern ) => {
-		if ( type ) {
+	const onSelect = ( type: string, selectedPattern: Pattern | null ) => {
+		if ( selectedPattern ) {
+			// Getselected pattern category or the first
+			selectedPattern.category = categories.find(
+				( { name } ) => name === ( categorySelected || selectedPattern.categories[ 0 ] )
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/index.tsx:280]: Could we make the `categorySelected` as a new parameter of this function? Then, this component doesn't need to handle the `categorySelected` state. Also for the `openPatternList` state. I think both could be encapsulated in the `ScreenCategoryList`. What do you think?

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/index.tsx [hunk 13]
```diff
@@ -428,12 +459,27 @@ const PatternAssembler: Step = ( { navigation, flow, stepName } ) => {
 					/>
 				</NavigatorScreen>
 				<NavigatorScreen path="/homepage/patterns">
-					<ScreenPatternList
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/index.tsx:431]: It would be better to keep this component if the feature flag is not enabled.

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/style.scss (hunk 3 above)
Comment (miksansegundo) [reply client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/style.scss:47]: Thanks. I think we can make this change without the flag.

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/navigator-buttons/style.scss [hunk 14]
```diff
@@ -3,12 +3,18 @@
...
 		font-size: $font-body-small;
 		font-weight: 500;
 		letter-spacing: -0.15px;
 		color: var(--studio-gray-100);
+		border-width: 2px;
+
+		&:focus {
+			border-radius: 3px;
+			border-color: var(--color-primary-light);
+		}
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/navigator-buttons/style.scss:17]: We can just raise another PR to update these styles 🙂

**Comment by miksansegundo** [comment]
I'm closing this in favor of https://github.com/Automattic/wp-calypso/pull/74149 and https://github.com/Automattic/wp-calypso/pull/74157. 

I'll continue working on more features in follow-up PRs.

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.scss [hunk 15]
```diff
@@ -1,12 +1,17 @@
...
-	margin-inline-start: 32px;
-	margin-top: -60px;
+	height: calc(100vh - 42px);
+	margin-inline-end: 32px;
+	margin-top: -50px;
 	box-sizing: border-box;
 	z-index: 1;
+
+	.device-switcher__frame {
+		box-shadow: none;
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.scss:11]: Why do we remove the box-shadow?

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.scss [hunk 16]
```diff
@@ -1,12 +1,17 @@
...
+		border-radius: 20px;
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.scss:12]: If we set the border-radius to `20px`, the border of the mobile device would be overwritten as well. I don't think we have to change the border of the mobile device here as it's also used in other places.

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/style.scss [hunk 17]
```diff
@@ -23,24 +31,26 @@ $font-family: "SF Pro Text", $sans;
 	.step-container__content {
 		display: flex;
 		width: 100%;
-		margin: 0 32px;
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/style.scss:26]: Why do we want to remove this?

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.scss [hunk 18]
```diff
@@ -1,12 +1,17 @@
 .pattern-large-preview {
 	flex: 1;
 	position: relative;
-	height: 100vh;
-	padding: 32px 0;
-	margin-inline-start: 32px;
-	margin-top: -60px;
+	height: calc(100vh - 42px);
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.scss:4]: Where is the `42px` from? If we want to make it higher, maybe changing the padding to something like `12px 0 32px` is enough. What do you think?

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/index.tsx [hunk 19]
```diff
@@ -373,36 +407,34 @@ const PatternAssembler: Step = ( { navigation, flow, stepName } ) => {
 		moveDownSection( position );
 	};
 
+	const handleClosePatternList = ( event: React.MouseEvent ) => {
+		// Click on large preview to close Pattern List
+		if ( ( event.target as HTMLElement ).closest( '.pattern-large-preview' ) ) {
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/index.tsx:412]: We can use `useFocusOutside` to handle this behavior. That is, close the pattern list when click outside the `ScreenCategoryList`

```js
import { __experimentalUseFocusOutside as useFocusOutside } from '@wordpress/compose';
```

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-layout.tsx [hunk 20]
```diff
@@ -24,125 +23,22 @@ type PatternLayoutProps = {
...
-				</AsyncLoad>
-			</div>
-		);
-	}
 
 	return (
 		<div className="pattern-layout">
 			{ sections.length > 0 && (
-				<AsyncLoad require="./animate-list" featureName="domMax" placeholder={ <div /> }>
+				<AsyncLoad require="./animate-presence" placeholder={ <div /> }>
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-layout.tsx:41]: Is there any issue to keep using orginal way?

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/hooks/use-pattern-categories-from-api.ts [hunk 21]
```diff
@@ -0,0 +1,32 @@
+import { useQuery, UseQueryResult, UseQueryOptions } from 'react-query';
+import wpcomRequest from 'wpcom-proxy-request';
+
+const usePatternCategoriesFromAPI = (
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/hooks/use-pattern-categories-from-api.ts:4]: [nit] I think `usePatternCategories` is enough 🙂

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/hooks/use-sections-map-by-category.ts [hunk 22]
```diff
@@ -0,0 +1,21 @@
+import { useMemo } from 'react';
+import type { Pattern, Category } from '../types';
+
+const useSectionsMapByCategory = ( patterns: Pattern[], categories: Category[] ) => {
+	return useMemo(
+		() =>
+			categories.reduce( ( categoriesMap: { [ key: string ]: Pattern[] }, category ) => {
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/hooks/use-sections-map-by-category.ts:7]: If we traverse `patterns` inside `categories.reduce`, we will go through the pattern list sooo many times even if this calculation might be memo forever.

Instead, we can create the `categoriesMap` first, and then traverse `patterns` to push it to its categories. For example,

```js
const categoriesMap = categories.reduce( ( acc, cur ) => ( {
  ...acc,
  [ cur.name ]: [],
} );

patterns.forEach( pattern => {
  patterns.categories.forEach( category => {
    if ( categoriesMap[ category ] ) {
      categoriesMap[ category ].push( pattern );
    }
  } );
} );
```

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/animate-with-features.tsx [hunk 23]
```diff
@@ -0,0 +1,16 @@
...
+	children: ( m: any ) => JSX.Element | null;
+}
+
+const AnimatePresenceWithFeatures = ( { featureName, children }: Props ) => (
+	<LazyMotion features={ () => loadFramerFeatures( featureName ) } strict>
+		{ children( m ) }
+	</LazyMotion>
+);
+
+export default AnimatePresenceWithFeatures;
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/animate-with-features.tsx:16]: Could we align the component name to the file name?

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-list.tsx [hunk 24]
```diff
@@ -1,57 +1,73 @@
...
-						title={ pattern.category }
-						className={ classnames( 'pattern-list__pattern', {
-							[ activeClassName ]: pattern.id === selectedPattern?.id,
-						} ) }
-						onClick={ () => onSelect( pattern ) }
-					/>
-				</PatternPreviewAutoHeight>
-			) ) }
-		</>
+		<AnimateWithFeatures featureName="domAnimation">
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-list.tsx:44]: If we do it this way, we will load the `framer-motion` when user lands on the pattern assembler, and that's why we did it via `AsyncLoad`. You can see the size changes a lot (maybe not just related to this one)

```
with-theme-assembler-flow     +899551 B  (+393.3%)  +267870 B  (+396.8%)
import-flow                   +868420 B   (+46.6%)  +257083 B   (+55.7%)
site-setup-flow               +832975 B   (+42.2%)  +244793 B   (+49.5%)
```

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/hooks/use-categories-order.ts [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+	'posts',
+	'text',
+	'gallery',
+	'images',
+	'call-to-action',
+	'banner',
+];
+
+const useCategoriesOrder = ( categories: Category[] ) => {
+	return useMemo(
```
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/hooks/use-categories-order.ts:15]: [nit] It might be slightly simple if we split `categories` into 2 arrays, and combine the result
* Categories with the pre-defined order
* The others

**Comment by arthur791004** [comment]
> I'll continue working on more features in follow-up PRs.

We can also raise some PRs to update the styles or related changes first.

**Comment by miksansegundo** [comment]
@arthur791004 thanks for your review, I'll address the feedback in https://github.com/Automattic/wp-calypso/pull/74157
